sdk/core/System.ClientModel/src/Convenience/ClientResultException.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.ClientModel.Internal;
+using System.ClientModel.Primitives;
+using System.Globalization;
+using System.Runtime.Serialization;
+using System.Text;
+
+namespace System.ClientModel;
+
+[Serializable]
+public class ClientResultException : Exception, ISerializable
+{
+    private const string DefaultMessage = "Service request failed.";
+
+    private readonly PipelineResponse? _response;
+    private int _status;
+
+    /// <summary>
+    /// Gets the HTTP status code of the response. Returns. <code>0</code> if response was not received.
+    /// </summary>
+    public int Status
+    {
+        get => _status;
+        protected set => _status = value;
+    }
+
+    public ClientResultException(PipelineResponse response, Exception? innerException = default)
+        : base(GetMessage(response), innerException)
+    {
+        ClientUtilities.AssertNotNull(response, nameof(response));
+
+        _response = response;
+        _status = response.Status;
+    }
+
+    public ClientResultException(string message, PipelineResponse? response = default, Exception? innerException = default)
+        : base(message ?? DefaultMessage, innerException)
+    {
+        _response = response;
+        _status = response?.Status ?? 0;
+    }
+
+    /// <summary>
+    /// TBD
+    /// </summary>
+    /// <param name="info"></param>
+    /// <param name="context"></param>
+    protected ClientResultException(SerializationInfo info, StreamingContext context)
+        : base(info, context)
+    {
+        _status = info.GetInt32(nameof(Status));
+    }
+
+    /// <inheritdoc />
+    public override void GetObjectData(SerializationInfo info, StreamingContext context)
+    {
+        ClientUtilities.AssertNotNull(info, nameof(info));
+
+        info.AddValue(nameof(Status), Status);
+
+        base.GetObjectData(info, context);
+    }
+
+    public PipelineResponse? GetRawResponse() => _response;
+
+    private static string GetMessage(PipelineResponse response)
```

Comment:
It may be helpful to call this something like `GenerateMessage` to indicate that it does a non-trivial amount of work for each invocation.

Reply:
Happy to change this.  I'll update the corresponding Azure.Core type as well in the Azure.Core 2.0 feature branch to make that change across the board.